### PR TITLE
Watch cache refactor watch cache store and add tests

### PR DESCRIPTION
**Pull request by serathius**
/kind cleanup

```release-note
NONE
```

Refactor for https://github.com/kubernetes/kubernetes/pull/126754

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go [hunk 1]
```diff
@@ -550,10 +479,29 @@ func (w *watchCache) waitUntilFreshAndListItems(ctx context.Context, resourceVer
 		}
 		return w.store.List(), w.resourceVersion, "", nil
 	}()
-
+	result, err = filterPrefix(key, result)
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go:482]: Adding it here means that both filtering and sorting will happen under lock, which we don't want.

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go:482]: I suggest leaving the `WaitUntilFreshAndList` as is just factor out the filter function as you did.

Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go]: Nice catch. Forgot about deferred lock.

**Comment by wojtek-t** [comment]
gofmt is failing - other than that LGTM

**Comment by serathius** [comment]
Done @wojtek-t

**Pull request merged**